### DISCUSSION TITLE
OpenAPI Generated for NamedTuple is Object, Not List

### DISCUSSION THREAD

**Discussion by michaeligreenberg** [fastapi/fastapi#9310] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import NamedTuple, Union

from fastapi import FastAPI
from pydantic import BaseModel

class Location(NamedTuple):
    latitude: float
    longitude: float

class Trip(BaseModel):
    start: Location
    end: Location

app = FastAPI()

@app.post("/trip")
def create_trip(trip: Trip):
    return {"status": "OK"}
```

### Description

* The `Location` NamedTuple provides constraints and a better interface for accessing a [lat,long] coordinate array.
* The /openapi.json endpoint produces the following schema definition of Location:
```
"Location":
{
    "title": "Location",
    "required":
    [
        "latitude",
        "longitude"
    ],
    "type": "object",
    "properties":
    {
        "latitude":
        {
            "title": "Latitude",
            "type": "number"
        },
        "longitude":
        {
            "title": "Longitude",
            "type": "number"
        }
    }
}
```
Note that it defines Location to be an object with named properties; this is incompatible with what the /trip endpoint actually accepts.
* Expected behavior is to define Location as a collection of items with minimum size 2, maximum size 2, and both entires of type number.

### Operating System

Linux

### Operating System Details

ubuntu 20.04 run via Docker

### FastAPI Version

0.95.0

### Python Version

3.8.10

**Comment by caniko** [comment]
In version 0.112 I get:
```openapi
    UploadBenchmarkingData:
      prefixItems:
      - type: string
        format: duration
        title: Upload Interval
      - type: integer
        exclusiveMinimum: 0.0
        title: Payload Byte Size
      - type: string
        title: Continent
      - type: string
        pattern: ^\w{2}$
        title: Country
      - type: string
        title: City
      type: array
      maxItems: 5
      minItems: 5
```
For:
```python
UploadBenchmarkingData = NamedTuple(
    "UploadBenchmarkingData",
    [
        ("upload_interval", datetime.timedelta),
        ("payload_byte_size", PositiveInt),
        ("continent", str),
        ("country", CountryAlpha2),
        ("city", str),
    ],
)
```

This is invalid, arrays are not supposed to have prefixItems, not sure what the right field is though.

**Comment by MythicalCosmic** [comment]
The issue is that NamedTuple doesn't work well with FastAPI/Pydantic because of a schema mismatch.  Use Pydantic BaseModel.

**Comment by CodeKraken-cmd** [comment]
FastAPI and Pydantic treat typing.NamedTuple as an object with named fields, so the generated schema is an object with latitude and longitude. That is expected. If your endpoint should accept a two-element array like [lat, lon], model it as a tuple or a constrained list instead of a NamedTuple. On newer FastAPI with OpenAPI 3.1, tuples appear in JSON Schema using prefixItems. That is valid JSON Schema 2020-12, but some tools that only understand OpenAPI 3.0 may not handle it. You have three practical choices:

1. Use a fixed-length tuple if the client can handle OpenAPI 3.1
2. Use a constrained list if you need 3.0-compatible tooling
3. Keep an object model if you prefer explicit field names and add range validation

Option A. Two-element numeric tuple (OpenAPI 3.1 will use prefixItems)
```
from typing import Tuple
from fastapi import FastAPI
from pydantic import BaseModel

Coord = Tuple[float, float]

class Trip(BaseModel):
    start: Coord
    end: Coord

app = FastAPI()  # emits OpenAPI 3.1 by default in recent versions

@app.post("/trip")
def create_trip(trip: Trip):
    return {"status": "OK"}

```

Option B. Constrained list with exact length 2 (plays nicely with OpenAPI 3.0)
```
from fastapi import FastAPI
from pydantic import BaseModel, conlist

Coord = conlist(float, min_length=2, max_length=2)

class Trip(BaseModel):
    start: Coord
    end: Coord

app = FastAPI(openapi_version="3.0.3")  # optional if you want to avoid prefixItems

@app.post("/trip")
def create_trip(trip: Trip):
    return {"status": "OK"}
```

Option C. Keep object shape with validation (explicit names and ranges)
```
from fastapi import FastAPI
from pydantic import BaseModel, Field

class Location(BaseModel):
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)

class Trip(BaseModel):
    start: Location
    end: Location

app = FastAPI()

@app.post("/trip")
def create_trip(trip: Trip):
    return {"status": "OK"}

```
